plugins/org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/annotations/validation/LinkingDiagnosticTypeAwareMessageProducer.xtend

```diff
@@ -0,0 +1,123 @@
+/** 
+ * Copyright (c) 2015 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.xtext.xbase.annotations.validation
+
+import com.google.inject.Inject
+import org.eclipse.emf.ecore.EClass
+import org.eclipse.emf.ecore.EObject
+import org.eclipse.emf.ecore.EStructuralFeature
+import org.eclipse.xtext.EcoreUtil2
+import org.eclipse.xtext.common.types.TypesPackage
+import org.eclipse.xtext.diagnostics.DiagnosticMessage
+import org.eclipse.xtext.linking.impl.IllegalNodeException
+import org.eclipse.xtext.linking.impl.LinkingDiagnosticMessageProvider
+import org.eclipse.xtext.xbase.XAbstractFeatureCall
+import org.eclipse.xtext.xbase.XFeatureCall
+import org.eclipse.xtext.xbase.XMemberFeatureCall
+import org.eclipse.xtext.xbase.annotations.xAnnotations.XAnnotationsPackage
+import org.eclipse.xtext.xbase.typesystem.internal.FeatureLinkHelper
+import org.eclipse.xtext.xbase.typesystem.internal.TypeAwareLinkingDiagnosticContext
+import org.eclipse.xtext.xbase.typesystem.references.LightweightTypeReference
+import org.eclipse.xtext.xbase.util.FeatureCallAsTypeLiteralHelper
+
+import static org.eclipse.xtext.diagnostics.Diagnostic.LINKING_DIAGNOSTIC
+import static org.eclipse.xtext.diagnostics.Severity.ERROR
+import static org.eclipse.xtext.xbase.XbasePackage.Literals.XMEMBER_FEATURE_CALL__MEMBER_CALL_TARGET
+
+/** 
+ * @author Dennis Huebner - Initial contribution and API
+ */
+class LinkingDiagnosticTypeAwareMessageProducer extends LinkingDiagnosticMessageProvider {
```

Comment:
This class name reads a little be weird. Can this be TypeAwareLinkingDiagnosticMessageProducer ?